src/test/java/tw/commonground/backend/service/fact/FactServiceTest.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    @Test
+    void testGetUrlDetails_withIconInLink() throws Exception {
+        Document mockDocument = new Document("https://www.github.com");
+        Element head = mockDocument.appendElement("head");
+        mockDocument.title("Github");
+        head.appendElement("link")
+                .attr("rel", "icon")
+                .attr("href", "https://www.github.com/favicon.ico");
+
+        FactService spyFactService = Mockito.spy(new FactService(factRepository, referenceRepository));
+        when(spyFactService.getDocument("https://www.github.com")).thenReturn(mockDocument);
+
+        ReferenceEntity referenceEntity = spyFactService.getUrlDetails("https://www.github.com");
+        assertThat(referenceEntity.getUrl()).isEqualTo("https://www.github.com");
+        assertThat(referenceEntity.getTitle()).isEqualTo("Github");
+        assertThat(referenceEntity.getFavicon()).isEqualTo("https://www.github.com/favicon.ico");
+    }
+
+    @Test
+    void testGetUrlDetails_withIconInMeta() throws Exception {
+        Document mockDocument = new Document("https://www.github.com");
+        Element head = mockDocument.appendElement("head");
+        mockDocument.title("Github");
+        head.appendElement("meta")
+                .attr("itemprop", "image")
+                .attr("content", "/favicon.ico");
+
+        FactService spyFactService = Mockito.spy(new FactService(factRepository, referenceRepository));
+        when(spyFactService.getDocument("https://www.github.com")).thenReturn(mockDocument);
+
+        ReferenceEntity referenceEntity = spyFactService.getUrlDetails("https://www.github.com");
+        assertThat(referenceEntity.getUrl()).isEqualTo("https://www.github.com");
+        assertThat(referenceEntity.getTitle()).isEqualTo("Github");
+        assertThat(referenceEntity.getFavicon()).isEqualTo("https://www.github.com/favicon.ico");
+    }
+
+    @Test
+    void testUrlHandling() {
+        List<String> input = List.of("https://www.google.com");
+
+        List<String> result = factService.urlHandling(input);
+
+        List<String> expected = List.of("https://www.google.com");
+
+        assertThat(result).isEqualTo(expected);
+    }
+
+    @Test
+    void testUrlHandling_withUrlWithoutProtocol() {
+        List<String> input = List.of("www.google.com");
+
+        List<String> result = factService.urlHandling(input);
+
+        List<String> expected = List.of("https://www.google.com");
+
+        assertThat(result).isEqualTo(expected);
+    }
+
+    @Test
+    void testUrlHandling_withEncodedUrl() {
+        List<String> input = List.of("https%3A%2F%2Fwww.google.com");
+
+        List<String> result = factService.urlHandling(input);
+
+        List<String> expected = List.of("https://www.google.com");
+
+        assertThat(result).isEqualTo(expected);
+    }
+
+    @Test
+    void testUrlHandling_withEncodedUrlWithoutProtocol() {
+        List<String> input = List.of("https%3A%2F%2Fwww.google.com");
```

Comment:
Is the input here (withEncodedUrlWithoutProtocol but it includes https) incorrect?